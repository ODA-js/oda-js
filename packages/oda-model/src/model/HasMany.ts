import { Record } from 'immutable';

import { IEntityRef } from '../interfaces/IEntityRef';
import { IHasMany, IHasManyInit, IHasManyStore, IRelationTransform } from '../interfaces/IHasMany';
import { Relation } from './Relation';
import { TransformField, TransformRef } from './utils';
import { IFieldContext } from '../contexts/IFieldContext';

// tslint:disable-next-line:variable-name
export const DefaultHasMany: IHasManyStore = {
  name: null,
  title: null,
  description: null,
  fields: null,
  opposite: null,
  hasMany: null,
  //storage
  single: false,
  stored: false,
  embedded: false,
  // name
  fullName: null,
  normalName: null,
  shortName: null,
};

// tslint:disable-next-line:variable-name
export const HasManyTransform: IRelationTransform = {
  hasMany: TransformRef(),
  fields: TransformField(),
};

// tslint:disable-next-line:variable-name
export const HasManyStorage = Record(DefaultHasMany);

export class HasMany extends Relation<IHasManyInit, IHasManyStore> implements IHasMany {
  public get verb(): 'HasMany' {
    return 'HasMany';
  }
  public get ref(): IEntityRef {
    return this.store.get('hasMany', null);
  }
  public get hasMany(): IEntityRef {
    return this.store.get('hasMany', null);
  }
  protected transform(input: Partial<IHasManyInit>): IHasManyStore {
    const result: IHasManyStore = {} as any;
    if (input) {
      for (let f in input) {
        if (input.hasOwnProperty(f)) {
          if (f === 'hasMany') {
            result.hasMany = HasManyTransform.hasMany.transform(input.hasMany);
          } else if (f === 'fields') {
            result.fields = HasManyTransform.fields.transform(input.fields);
          } else {
            result[f] = input[f];
          }
        }
      }
    }
    return result;
  }

  protected reverse(input: Record<IHasManyStore> & Readonly<IHasManyStore>): IHasManyInit {
    const result: IHasManyInit = {} as any;
    if (input) {
      const core = input.toJS();
      for (let f in core) {
        if (core.hasOwnProperty(f)) {
          if (f === 'hasMany' || f === 'ref') {
            result.hasMany = HasManyTransform.hasMany.reverse(input.hasMany);
          } else if (f === 'fields') {
            result.fields = HasManyTransform.fields.reverse(input.fields);
          } else {
            result[f] = core[f];
          }
        }
      }
    }
    return result;
  }
  constructor(init: Partial<IHasManyInit>, context: IFieldContext) {
    super(init, context);
    this.store = new HasManyStorage(this.transform(init));
    this.init = new (Record<Partial<IHasManyInit>>(init))();
  }
}
