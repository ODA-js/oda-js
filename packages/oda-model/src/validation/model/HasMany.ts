import { IRelationPropsStore } from '../interfaces/IRelation';
import { Record } from 'immutable';
import { Map, Set } from 'immutable';

import { Persistent } from './Persistent';
import { transformMap, transformSet } from './utils';
import { IHasManyRelationPropsStore, IHasManyRelationProps, IHasManyRelation } from '../interfaces/IHasManyRelation';
import { IEntityRef } from '../interfaces/IEntityRef';
import { IField } from '../interfaces/IField';

// tslint:disable-next-line:variable-name
export const DefaultHasMany: IHasManyRelationPropsStore = {
  modelType: 'relation',
  name: null,
  title: null,
  description: null,
  verb: null,
  ref: null,
  fields: null,
  opposite: null,
  hasMany: null,
  //storage
  single: true,
  stored: true,
  embedded: true,
// name
  fullName: null,
  normalName: null,
  shortName: null,
};

// tslint:disable-next-line:variable-name
export const RelationTransform: { [ k in keyof IHasManyRelationPropsStore]?: any } = {
  fields: transformMap<IField>(),
};

// tslint:disable-next-line:variable-name
export const HasManyStorage = Record(DefaultHasMany);

export class HasMany extends Persistent<IHasManyRelationProps, IHasManyRelationPropsStore> implements IHasManyRelation {
  public get modelType(): 'relation' {
    return 'relation';
  }
  public get name(): string {
    return this.store.get('name', null);
  }
  public get title(): string {
    return this.store.get('title', null);
  }
  public get description(): string {
    return this.store.get('description', null);
  }
  public get ref(): IEntityRef {
    return this.store.get('ref', null);
  }
  public get fields(): Map<string, IField> {
    return this.store.get('fields', null);
  }
  public get opposite(): string {
    return this.store.get('opposite', null);
  }
  public get embedded(): boolean {
    return this.store.get('embedded', null);
  }
  public get single(): boolean {
    return this.store.get('single', null);
  }
  public get stored(): boolean {
    return this.store.get('stored', null);
  }
  public get fullName(): string {
    return this.store.get('fullName', null);
  }
  public get normalName(): string {
    return this.store.get('normalName', null);
  }
  public get shortName(): string {
    return this.store.get('shortName', null);
  }
  public get verb(): 'HasMany' {
    return 'HasMany';
  }
  public get hasMany(): IEntityRef {
    return this.store.get('hasMany', null);
  }

  protected transform(input: IHasManyRelationProps): IHasManyRelationPropsStore {
    return {
      ...input,
      single: true,
      stored: true,
      embedded: true,
      verb: 'HasMany',
      fields: RelationTransform.fields.transform(input.fields),
    };
  }
  protected reverse(input: IHasManyRelationPropsStore): IHasManyRelationProps {
    return {
      name: input.name,
      title: input.title,
      description: input.description,
      modelType: input.modelType,
      fullName: input.fullName,
      normalName: input.normalName,
      shortName: input.shortName,
      opposite: input.opposite,
      ref: input.ref,
      verb: input.verb,
      hasMany: input.hasMany,
      fields: RelationTransform.fields.reverse(input.fields),
    };
  }
  constructor(init: IHasManyRelationProps) {
    super();
    this.store = new HasManyStorage(this.transform(init));
    this.init = new (Record<IHasManyRelationProps>(init))();
  }
}
