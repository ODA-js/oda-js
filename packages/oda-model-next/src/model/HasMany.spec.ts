import 'jest';
import { HasMany } from './HasMany';

describe('HasMany', () => {
  let relation: HasMany;

  beforeAll(() => {
    expect(() => new HasMany()).not.toThrow();
  });

  beforeEach(() => {
    relation = new HasMany();
  });

  it('create empty', () => {
    expect(relation).toMatchSnapshot();
    expect(relation.toJS()).toMatchSnapshot();
  });

  it('updates strings', () => {
    expect(() =>
      relation.updateWith({
        name: 'cool',
        description: 'very cool',
        title: 'very cool title',
      }),
    ).not.toThrow();
    expect(relation.name).toBe('cool');
    expect(relation.description).toBe('very cool');
    expect(relation.title).toBe('very cool title');
    expect(() =>
      relation.updateWith({
        name: 'cool!',
      }),
    ).not.toThrow();
    expect(relation.name).toBe('cool!');
  });

  it('updates with null or undefined', () => {
    expect(() =>
      relation.updateWith({
        name: 'cool',
        description: 'very cool',
        title: 'very cool title',
      }),
    ).not.toThrow();
    expect(relation.toJS()).toMatchSnapshot();
    expect(() =>
      relation.updateWith({
        description: null,
        title: undefined,
      }),
    ).not.toThrow();
    expect(relation.toJS()).toMatchSnapshot();
  });

  it('toJS with dupes', () => {
    expect(() =>
      relation.updateWith({
        fields: [{ name: 'one' }, { name: 'one' }],
      }),
    ).not.toThrow();

    expect(relation.toJS()).toMatchSnapshot();
  });

  it('toJS with HashMap', () => {
    expect(() =>
      relation.updateWith({
        fields: {
          one: {},
        },
      }),
    ).not.toThrow();

    expect(relation.toJS()).toMatchSnapshot();
  });
  it('toJS ', () => {
    expect(() =>
      relation.updateWith({
        hasMany: 'id@Some#id',
      }),
    ).not.toThrow();

    expect(relation.ref.toJS()).toMatchSnapshot();
    expect(relation.toJS()).toMatchSnapshot();
  });
});
