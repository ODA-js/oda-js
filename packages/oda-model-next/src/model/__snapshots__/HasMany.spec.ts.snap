// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HasMany create empty 1`] = `
HasMany {
  "store": Immutable.Record {
    "name": null,
    "title": null,
    "description": null,
    "fields": null,
    "opposite": null,
    "hasMany": null,
    "single": false,
    "stored": false,
    "embedded": false,
    "fullName": null,
    "normalName": null,
    "shortName": null,
  },
}
`;

exports[`HasMany create empty 2`] = `
Object {
  "embedded": false,
  "single": false,
  "stored": false,
}
`;

exports[`HasMany toJS  1`] = `
Object {
  "backField": "id",
  "entity": "Some",
  "field": "id",
}
`;

exports[`HasMany toJS  2`] = `
Object {
  "embedded": false,
  "hasMany": "id@Some#id",
  "single": false,
  "stored": false,
}
`;

exports[`HasMany toJS with HashMap 1`] = `
Object {
  "embedded": false,
  "fields": Array [
    Object {
      "name": "one",
    },
  ],
  "single": false,
  "stored": false,
}
`;

exports[`HasMany toJS with dupes 1`] = `
Object {
  "embedded": false,
  "fields": Array [
    Object {
      "name": "one",
    },
  ],
  "single": false,
  "stored": false,
}
`;

exports[`HasMany updates with null or undefined 1`] = `
Object {
  "description": "very cool",
  "embedded": false,
  "name": "cool",
  "single": false,
  "stored": false,
  "title": "very cool title",
}
`;

exports[`HasMany updates with null or undefined 2`] = `
Object {
  "embedded": false,
  "name": "cool",
  "single": false,
  "stored": false,
}
`;
